crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -525,45 +555,175 @@ impl<'w, 's> SystemParam for &'w World {
+            .add_read(ArchetypeComponentId::WORLD_METADATA);
+        // prevent executor from concurrently running any systems that mutably borrow data
+        system_meta.archetype_component_access.read_all();
+
+        WorldState
+    }
+
+    // fn new_archetype(&mut self, archetype: &Archetype, system_meta: &mut SystemMeta) {
+    //     for info in archetype.components.values() {
+    //         system_meta.archetype_component_access.add_read(info.archetype_component_id);
+    //     }
+    // }
+}
+
+impl<'w, 's> SystemParamFetch<'w, 's> for WorldState {
+    type Item = &'w World;
+    unsafe fn get_param(
+        _state: &'s mut Self,
+        _system_meta: &SystemMeta,
+        world: &SemiSafeCell<'w, World>,
+        _change_tick: u32,
+    ) -> Self::Item {
+        world.as_ref()
+    }
+}
+
+/// SAFETY: &World is an immutable borrow.
+unsafe impl ReadOnlySystemParamFetch for WorldState {}
+
+/// The [`SystemParamState`] of [`&mut World`](crate::world::World).
+pub struct WorldMutState;
+
+impl<'w, 's> SystemParam for &'w mut World {
+    type Fetch = WorldMutState;
+}
+
+unsafe impl<'w, 's> SystemParamState for WorldMutState {
+    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self {
+        // world could contain non-send resources, run on local thread
+        system_meta.set_non_send();
+
+        let mut component_access = FilteredAccess::default();
+        let id = world.init_component::<WorldMetadata>();
+        component_access.add_write(id);
+        component_access.write_all();
+
+        if !system_meta
+            .component_access_set
+            .get_conflicts(&component_access)
+            .is_empty()
         {
-            panic!("&World conflicts with a previous mutable system parameter. Allowing this would break Rust's mutability rules");
+            panic!(
+                "&mut World conflicts with another system param in {}. \
+                Mutable access must be unique.",
+                system_meta.name,
+            );
         }
-        system_meta.archetype_component_access.extend(&access);
 
-        let mut filtered_access = FilteredAccess::default();
+        // this is temporary, see #4166
+        system_meta.set_exclusive();
+
+        system_meta.component_access_set.add(component_access);
+        // make sure executor sees conflict with another system reading WorldMetadata
+        system_meta
+            .archetype_component_access
+            .add_write(ArchetypeComponentId::WORLD_METADATA);
+        // prevent executor from concurrently running other systems that borrow data
+        system_meta.archetype_component_access.write_all();
+
+        WorldMutState
+    }
+
+    // fn new_archetype(&mut self, archetype: &Archetype, system_meta: &mut SystemMeta) {
+    //     for info in archetype.components.values() {
+    //         system_meta.archetype_component_access.add_write(info.archetype_component_id);
+    //     }
+    // }
```

Comment:
Alternatively, we can avoid these confusing set theory questions by removing `reads_all` and `writes_all` and having the `&mut World` set bits like `Query` does.